hyperspace/cosmos/src/lib.rs

```diff
@@ -11,3 +11,221 @@
+			path::ClientConsensusStatePath,
+			Path,
+		},
+	},
+	protobuf::Protobuf,
+};
+use ibc_proto::{
+	cosmos::{
+		auth::v1beta1::{query_client::QueryClient, BaseAccount, QueryAccountRequest},
+		base::query::v1beta1::PageRequest,
+	},
+	google::protobuf::Any,
+	ibc::core::{
+		client::v1::{
+			IdentifiedClientState, QueryConsensusStateRequest, QueryConsensusStateResponse,
+		},
+		connection::v1::{IdentifiedConnection, QueryConnectionResponse},
+	},
+};
+use ics07_tendermint::{
+	client_state::ClientState as TmClientState, consensus_state::ConsensusState as TmConsensusState,
+};
+use key_provider::KeyEntry;
+use pallet_ibc::{
+	light_clients::{AnyClientState, AnyConsensusState, HostFunctionsManager},
+	MultiAddress, Timeout, TransferParams,
+};
+use primitives::{IbcProvider, KeyProvider};
+use prost::Message;
+use serde::Deserialize;
+use std::{str::FromStr, sync::Arc, time::Duration};
+use tendermint::block::Height as TmHeight;
+use tendermint::time::Time;
+use tendermint_rpc::{abci::Path as TendermintABCIPath, Client, HttpClient, Url, WebSocketClient};
+use tendermint_verifier::LightClient;
+// Implements the [`crate::Chain`] trait for cosmos.
+/// This is responsible for:
+/// 1. Tracking a cosmos light client on a counter-party chain, advancing this light
+/// client state  as new finality proofs are observed.
+/// 2. Submiting new IBC messages to this cosmos.
+#[derive(Clone)]
+pub struct CosmosClient<H> {
+	/// Chain name
+	pub name: String,
+	/// Chain rpc client
+	pub rpc_client: HttpClient,
+	/// Chain grpc address
+	pub grpc_url: Url,
+	/// Websocket chain ws client
+	pub websocket_url: Url,
+	/// Chain Id
+	pub chain_id: ChainId,
+	/// Light client id on counterparty chain
+	pub client_id: Option<ClientId>,
+	/// Connection Id
+	pub connection_id: Option<ConnectionId>,
+	/// Light client to track the counterparty chain
+	pub light_client: LightClient,
+	/// Name of the key to use for signing
+	pub keybase: KeyEntry,
+	/// Account prefix
+	pub account_prefix: String,
+	/// Reference to commitment
+	pub commitment_prefix: CommitmentPrefix,
+	/// Channels cleared for packet relay
+	pub channel_whitelist: Vec<(ChannelId, PortId)>,
+	/// Finality protocol to use, eg Tenderminet
+	pub finality_protocol: finality_protocol::FinalityProtocol,
+	pub _phantom: std::marker::PhantomData<H>,
+}
+/// config options for [`ParachainClient`]
+#[derive(Debug, Deserialize)]
+pub struct CosmosClientConfig {
+	/// Chain name
+	pub name: String,
+	/// rpc url for cosmos
+	pub rpc_url: Url,
+	/// grpc url for cosmos
+	pub grpc_url: Url,
+	/// websocket url for cosmos
+	pub websocket_url: Url,
+	/// Cosmos chain Id
+	pub chain_id: String,
+	/// Light client id on counterparty chain
+	pub client_id: Option<String>,
+	/// Connection Id
+	pub connection_id: Option<String>,
+	/// Account prefix
+	pub account_prefix: String,
+	/// Store prefix
+	pub store_prefix: String,
+	/// Name of the key that signs transactions
+	pub key_name: String,
```

Comment:
there are some repeated fields in the CosmosClientConfig and CosmosClient struct. Can we define the `CosmosClientConfig` type in the `CosmosClient` struct and initialize in the `new` method?

Reply:
This isn't a real problem.

Reply:
what do you mean?

Reply:
There's nothing wrong with the duplicated fields across the config and client structs.